src/couch_index_monitor.erl

```diff
@@ -0,0 +1,184 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_index_monitor).
+
+
+-export([
+    spawn_link/2,
+    close/1,
+    set_pid/2,
+
+    notify/1,
+    notify/2,
+    cancel/2
+]).
+
+-export([
+    init/2
+]).
+
+
+-include("couch_index.hrl").
+
+
+-record(st, {
+    name,
+    type,
+    is_sys,
+    ref,
+    client_refs,
+    closing
+}).
+
+
+spawn_link(Name, IsSys) ->
+    erlang:spawn_link(?MODULE, init, [Name, IsSys]).
+
+
+close(Monitor) ->
+    Monitor ! exit,
+    ok.
+
+
+set_pid(Monitor, Pid) ->
+    Monitor ! {set_pid, Pid},
+    ok.
+
+
+notify(Monitor) ->
+    notify(Monitor, self()).
+
+
+notify(Monitor, Client) when is_pid(Client) ->
+    Monitor ! {notify, Client},
+    ok;
+
+notify(Monitor, {Client, _}) when is_pid(Client) ->
+    notify(Monitor, Client).
+
+
+cancel(Name, {Client, Monitor, IsSys})
+        when Client == self(), is_pid(Monitor) ->
+    Monitor ! {cancel, self()},
+    case (catch ets:update_counter(?BY_COUNTERS, Name, -1)) of
+        0 when not IsSys ->
+            true = ets:insert(?BY_IDLE, {Name}),
+            ok;
+        _ ->
+            ok
+    end.
+
+
+init(Name, IsSys) ->
+    {ok, CRefs} = khash:new(),
+    loop(#st{
+        name = Name,
+        is_sys = IsSys,
+        ref = undefined,
+        client_refs = CRefs,
+        closing = false
+    }).
+
+
+handle_info(exit, St) ->
```

Comment:
This looked like a gen_server at first but didn't see the exports and spawn looked different. Then noticed we did a custom loop. Is this to make it lightweight? Would a proc_lib process be better?